File: examples/printer_output.py
from emu_8085 import Data, Instruction, Machine, Mem, Opcode, Program, PrinterDevice


def main_run() -> None:
    # 1. Create a printer device with a custom logging callback
    printed_chars = []
    def log_printed_char(char: str):
        printed_chars.append(char)
        print(f"[Printer Output]: '{char}'")

    printer = PrinterDevice(output_callback=log_printed_char)

    # 2. Attach printer to port 0x05
    machine = Machine.create(
        address_lines=16,
        data_lines=8,
        devices=[(printer, [0x05])]
    )

    # 3. Create a program that outputs the letters 'A', 'B', 'C' to the printer
    program = Program([
        Instruction(Opcode.MVI_A, Data.byte(ord('A'))),
        Instruction(Opcode.OUT, Data.byte(0x05)),
        Instruction(Opcode.MVI_A, Data.byte(ord('B'))),
        Instruction(Opcode.OUT, Data.byte(0x05)),
        Instruction(Opcode.MVI_A, Data.byte(ord('C'))),
        Instruction(Opcode.OUT, Data.byte(0x05)),
        Instruction(Opcode.HLT)
    ])

    machine.load(program, Mem(0x0000))
    machine.run()

    print("\nPrinter Output Example State:")
    print(f"Printed character buffer: {printed_chars}")
    print(f"Printer device history: {printer.history}")

if __name__ == "__main__":
    main_run()

File: examples/printer_output.rs
//! Ported from the reference project's `examples/printer_output.py`.
//! Run with `cargo run --example printer_output`.

use emu8085::{Addr, Instruction, Machine, Opcode, Operand, PrinterDevice, Program};
use std::cell::RefCell;
use std::rc::Rc;

fn main() {
    // 1. Create a printer device with a custom logging callback
    let printed_chars = Rc::new(RefCell::new(Vec::new()));
    let sink = printed_chars.clone();
    let printer = PrinterDevice::with_callback(move |char| {
        sink.borrow_mut().push(char);
        println!("[Printer Output]: '{char}'");
    });

    // 2. Attach printer to port 0x05
    let mut machine = Machine::create(16, 8);
    machine.attach_device(Box::new(printer), &[0x05]);

    // 3. Create a program that outputs the letters 'A', 'B', 'C' to the printer
    let program = Program::new(vec![
        Instruction::with(Opcode::MVI_A, Operand::byte(b'A')),
        Instruction::with(Opcode::OUT, Operand::byte(0x05)),
        Instruction::with(Opcode::MVI_A, Operand::byte(b'B')),
        Instruction::with(Opcode::OUT, Operand::byte(0x05)),
        Instruction::with(Opcode::MVI_A, Operand::byte(b'C')),
        Instruction::with(Opcode::OUT, Operand::byte(0x05)),
        Instruction::new(Opcode::HLT),
    ]);

    machine.load(&program, Addr(0x0000)).expect("program compiles");
    machine.run();

    let term = machine.devices.device_ref::<PrinterDevice>(0).unwrap();
    println!("\nPrinter Output Example State:");
    println!("Printed character buffer: {:?}", printed_chars.borrow());
    println!("Printer device history: {}", term.history);

    assert_eq!(*printed_chars.borrow(), vec!['A', 'B', 'C']);
}
